nav2_costmap_2d/plugins/costmap_filters/keepout_filter.cpp

```diff
@@ -0,0 +1,166 @@
+}
+
+void KeepoutFilter::mapFilterCallback(
+  const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
+{
+  map_filter_ = msg;
+
+  // Updating costmap_filter_
+  unsigned int filter_cell_sz_x = map_filter_->info.width;
+  unsigned int filter_cell_sz_y = map_filter_->info.height;
+  double filter_map_res = map_filter_->info.resolution;
+  double filter_origin_x = map_filter_->info.origin.position.x;
+  double filter_origin_y = map_filter_->info.origin.position.y;
+  // ToDo - current Costmap2D does not support orientation
+  costmap_filter_ = std::make_unique<Costmap2D>(
+    filter_cell_sz_x, filter_cell_sz_y, filter_map_res, filter_origin_x, filter_origin_y);
+
+  // Fill the costmap_filter_ with values from map_filter_
+  unsigned int mx, my, it;
+  int8_t data;
+  for (mx = 0; mx < filter_cell_sz_x; mx++) {
+    for (my = 0; my < filter_cell_sz_y; my++) {
+      it = costmap_filter_->getIndex(mx, my);
+      data = map_filter_->data[it];
+      if (data == nav2_map_server::OCC_GRID_OCCUPIED) {
+        costmap_filter_->setCost(mx, my, LETHAL_OBSTACLE);
+      } else if (
+        data == nav2_map_server::OCC_GRID_FREE ||
+        data == nav2_map_server::OCC_GRID_UNKNOWN)
+      {
+        costmap_filter_->setCost(mx, my, NO_INFORMATION);
+      } else {
+        unsigned char cost = std::round(
+          data * (LETHAL_OBSTACLE - FREE_SPACE) / \
+          (nav2_map_server::OCC_GRID_OCCUPIED - nav2_map_server::OCC_GRID_FREE));
+        costmap_filter_->setCost(mx, my, cost);
+      }
+    }
+  }
+}
+
+void KeepoutFilter::process(
+  nav2_costmap_2d::Costmap2D & master_grid,
+  int min_i, int min_j, int max_i, int max_j,
+  double /*robot_x*/, double /*robot_y*/, double /*robot_yaw*/)
+{
+  if (!map_filter_) {
+    RCLCPP_WARN(node_->get_logger(), "Map filter was not received");
+    return;
+  }
+
+  unsigned int i, j;  // costmap_filter_ iterators
+  double wx, wy;  // world coordinates
+  unsigned int mx, my;  // costmap_ coordinates
+
+  unsigned char * costmap_filter_charmap = costmap_filter_->getCharMap();
+  unsigned char cost;
+
+  for (i = 0; i < costmap_filter_->getSizeInCellsX(); i++) {
+    for (j = 0; j < costmap_filter_->getSizeInCellsY(); j++) {
+      cost = costmap_filter_charmap[costmap_filter_->getIndex(i, j)];
+      if (cost != NO_INFORMATION) {
+        // Converting each point on costmap_filter to world coordinates
+        costmap_filter_->mapToWorld(i, j, wx, wy);
+        // Getting coordinates of costmap_ (if any) corresponding to given
+        // world coordinates
+        bool on_costmap = master_grid.worldToMap(wx, wy, mx, my);
+        // If these coordinates are belonging to costmap bounds put LETHAL_OBSTACLE
+        // on layer
+        if (on_costmap &&
+          mx >= (unsigned int)min_i && mx < (unsigned int)max_i &&
+          my >= (unsigned int)min_j && my < (unsigned int)max_j)
+        {
+          costmap_[master_grid.getIndex(mx, my)] = cost;
```

Comment:
There's no checking if anything is out of bounds, which would happen if you iterate outside the min/max

Reply:
I've moved `worldToMap()` call in the code above from `master_grid` to current costmap (`this`). So, it should never happen, because of in this case worldToMap() will return false value in each case when out of bounds.